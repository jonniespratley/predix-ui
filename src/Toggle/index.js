export { default } from './Toggle';
// export { ToggleLabel } from './Toggle';
