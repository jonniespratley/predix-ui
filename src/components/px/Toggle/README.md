The `<Toggle/>` is a simple on/off switch component. Use it to create UI elements that enable and disable things.


## Usage

```react
const { Toggle } = PxReact;
<div>
  <Toggle id="toggle1" name='new' value="true"/>
  <br/>
  <Toggle id="toggle2" name='used1' value="true" size='small'/>				
  <br/>
  <Toggle id="toggle3" name='used2' value="true" size='large'/>
  <br/>
  <Toggle id="toggle4" name='used3' value="true" size='huge'/>
  <br/>
  <Toggle id="toggle5" name='used4' value="true" size='huge' disabled/>
</div>
```


## Properties

```table
span: 6
rows:
  - Name: id
    Type: String
    Description: The id of the input
  - Name: name
    Type: String
    Description: The name of the input
  - Name: value
    Type: String
    Description: The value of the input
  - Name: checked
    Type: Boolean
    Description: The checked value of the input
  - Name: disabled
    Type: Boolean
    Description: The disabled value of the input
  
```


### Styling
The following custom properties are available for styling.

```table
span: 6
rows:
  - Property: --toggle-input
    Default: null
    Description: Mixin applied to entire element.
```
