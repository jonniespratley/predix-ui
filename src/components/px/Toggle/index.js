import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

const ToggleLabel = styled.label`
  position: relative;
  display: block;
  width: 2.66667rem;
  min-width: 2.66667rem;
  height: calc(1rem - 1px);
  margin-bottom: 0;
  margin-left: 0;
  padding: 1px;
  cursor: pointer;
  outline: 0;
  user-select: none;
  background-color: var(--px-toggle__background-border--unchecked, #889aa5);
  text-indent: -9999rem;
  pointer-events: auto;
  border-radius: 2rem;
  transition: background 0.4s;
  &:before {
    border-radius: 2rem;
    transition: background 0.4s;
  }
  &:after {
    position: absolute;
    display: block;
    content: "";
  }
  &:before {
    position: absolute;
    display: block;
    content: "";
    top: 1px;
    right: 1px;
    bottom: 1px;
    left: 1px;
    background-color: var(--px-toggle__background--unchecked, #d8e0e5);
  }
  &:hover {
    &:before {
      background-color: var(--px-toggle__background--unchecked--hover, #b6c3cc);
    }
    background-color: var(--px-toggle__background-border--unchecked--hover, #677e8c);
  }
  &:active {
    &:before {
      background-color: var(--px-toggle__background--unchecked--pressed, #748b99);
    }
    background-color: var(--px-toggle__background-border--unchecked-pressed, #4f6472);
  }
  &:after {
    top: -3px;
    left: -1px;
    height: 1.33333rem;
    width: 1.33333rem;
    border: 1px solid var(--px-toggle__switch-border, #889aa5);
    border-radius: 50%;
    box-shadow: 0 1px 2px var(--px-toggle__switch-shadow-color, rgba(0, 0, 0, 0.15));
    background-color: var(--px-toggle__switch, #fff);
    transition: margin 0.4s, background 0.4s, border-color 0.4s;
  }
  &:hover:after {
    background-color: var(--px-toggle__switch--hover, #c5d1d8);
    border: 1px solid var(--px-toggle__switch-border--hover, #889aa5);
  }
  &:active:after {
    background-color: var(--px-toggle__switch--pressed, #96a8b2);
    border: 1px solid var(--px-toggle__switch-border--pressed, #889aa5);
  }
  ${props => props.disabled && css`
    cursor: not-allowed;
    background-color: var(--px-toggle__background-border--unchecked--disabled, rgba(0, 0, 0, 0.2));
    pointer-events: none;
  `}
  ${props => props.size === 'huge' && css`
    width: 5.33333rem;
    min-width: 5.33333rem;
    height: 2rem;
    &:after {
      top: -.33333rem;
      left: -.06667rem;
      height: calc(2.66667rem);
      width: calc(2.66667rem);
    }
  `}
  ${props => props.size === 'large' && css`
    width: 4rem;
    min-width: 4rem;
    height: 1.33333rem;
    &:after {
      top: -.33333rem;
      left: -.06667rem;
      height: 2rem;
      width: 2rem;
    }
  `}
  ${props => props.size === 'small' && css`
    width: 2rem;
    min-width: 2rem;
    height: calc(.66667rem + 1px);
    &:after {
      top: -2px;
      left: 0;
      height: 1rem;
      width: 1rem;
    }
  `}
}`;
const ToggleInputStyle = styled.input`
  position: absolute !important;
  overflow: hidden !important;
  width: 1px !important;
  height: 1px !important;
  margin: -1px !important;
  clip: rect(0 0 0 0) !important;

  &:disabled,
  &:disabled + label {

    background-color: var(--px-toggle__background-border--unchecked--disabled, rgba(0, 0, 0, 0.2));
    pointer-events: none;

    &:after{
      background-color: var(--px-toggle__switch--disabled, #fff);
    }
    &:before{
        background-color: var(--px-toggle__background--unchecked--disabled, #fff);
        box-shadow: none;
    }
  }
  &:checked:disabled + label{
    background-color: var(--px-toggle__background--checked--disabled, rgba(0, 0, 0, 0.2));
    &:after{
      box-shadow: none;
      background-color: var(--px-toggle__switch--disabled, white);
      border-width: 1px;
      border-style: solid;
      border-color: var(--px-toggle__switch-border--disabled, rgba(0, 0, 0, 0.2));
    }
    &:before{
        background-color: transparent;
    }
  }
  &:checked + label {
    background-color: var(--px-toggle__background--checked, #007acc);
    &:before {
        background-color: var(--px-toggle__background--checked, #007acc);
    }
    &:hover {
        background-color: var(--px-toggle__background--checked--hover, #005c99);
        &:before {
            background-color: var(--px-toggle__background--checked--hover, #005c99);
        }
    }
    &:active {
        background-color: var(--px-toggle__background--checked--pressed, #003d66);
        &:before {
            background-color: var(--px-toggle__background--checked--pressed, #003d66);
        }
    }
    &:after {
        margin-left: 1.467rem;
    }
    ${props => props.size === 'huge' && css`
      width: 5.33333rem;
      min-width: 5.33333rem;
      height: 2rem;
      &:after {
        margin-left: 2.8rem;
        top: -.33333rem;
        left: -.06667rem;
        height: calc(2.66667rem);
        width: calc(2.66667rem);
      }
    `}
    ${props => props.size === 'large' && css`
      width: 4rem;
      min-width: 4rem;
      height: 1.33333rem;
      &:after {
        margin-left: 2.133rem;
        top: -.33333rem;
        left: -.06667rem;
        height: 2rem;
        width: 2rem;
      }
    `}
    ${props => props.size === 'small' && css`
    `}
  }
`;

const ToggleInput = ({
  checked,
  name,
  value,
  id,
  disabled,
  className,
  onClick,
  size
}) => (
  <div className={className}>
    <ToggleInputStyle
      id={id}
      size={size}
      name={name}
      value={value}
      type="checkbox"
      checked={checked}
      disabled={disabled}
    />
    <ToggleLabel
      disabled={disabled}
      size={size}
      onClick={onClick}
      for={id}
    />
  </div>
);

ToggleInput.displayName = 'ToggleInput';

ToggleInput.propTypes = {
  id: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  checked: PropTypes.bool,
  value: PropTypes.string,
  size: PropTypes.string,
  className: PropTypes.string,
  onClick: PropTypes.func,
  name: PropTypes.string
};

ToggleInput.defaultProps = {
  disabled: null,
  checked: null,
  size: null,
  className: null,
  value: null,
  onClick: null,
  name: null
};

export default ToggleInput;
