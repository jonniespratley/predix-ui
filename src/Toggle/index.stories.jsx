import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, boolean, select } from '@storybook/addon-knobs/react';
import Toggle from '.';

const sizeOptions = {
  small: 'Small',
  regular: 'Regular',
  large: 'Large',
  huge: 'Huge'
};

const stories = storiesOf('Components / Toggle', module);
stories.addDecorator(withKnobs);

stories.addWithJSX('default', () => (
  <form>
    <Toggle
      id="toggle2"
      name="used1"
      value="true"
      size={select('size', sizeOptions, 'regular')}
      disabled={boolean('disabled', false)}
      defaultChecked={boolean('checked', true)}
    />
  </form>
));
stories.addWithJSX('controlled', () => (
  <form>
    <Toggle
      id="toggle2"
      name="used1"
      value="true"
      size={select('size', sizeOptions, 'regular')}
      onChange={action('onChange')}
      disabled={boolean('disabled', false)}
      checked={boolean('checked', true)}
    />
  </form>
));
