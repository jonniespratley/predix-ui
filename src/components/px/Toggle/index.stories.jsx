import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, select } from '@storybook/addon-knobs/react';
import Toggle from './';

const sizeOptions = {
  small: 'Small',
  regular: 'Regular',
  large: 'Large',
  huge: 'Huge'
};

const stories = storiesOf('Components / Toggle', module);
stories.addDecorator(withKnobs);

stories.addWithJSX('default', () => (
  <form>
    <Toggle
      id="toggle2"
      name="used1"
      value="true"
      size={select('size', sizeOptions, 'regular')}
      disabled={boolean('disabled', false)}
      checked={boolean('checked', false)}
    />
  </form>
));
