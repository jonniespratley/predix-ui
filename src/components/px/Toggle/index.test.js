import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Toggle from './';

describe('Toggle', () => {

  test('matches snapshot', () => {
    const tree = renderer.create(<Toggle name='toggle1' id='toggle1' />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('disabled = matches snapshot', () => {
    const tree = renderer.create(<Toggle name='toggle1' id='toggle1' disabled/>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
