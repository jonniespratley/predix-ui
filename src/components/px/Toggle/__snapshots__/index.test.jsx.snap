// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle disabled = matches snapshot 1`] = `
.c1 {
  position: relative;
  display: block;
  width: 2.66667rem;
  min-width: 2.66667rem;
  height: calc(1rem - 1px);
  margin-bottom: 0;
  margin-left: 0;
  padding: 1px;
  cursor: pointer;
  outline: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: var(--px-toggle__background-border--unchecked,#889aa5);
  text-indent: -9999rem;
  pointer-events: auto;
  border-radius: 2rem;
  -webkit-transition: background 0.4s;
  transition: background 0.4s;
  cursor: not-allowed;
  background-color: var(--px-toggle__background-border--unchecked--disabled,rgba(0,0,0,0.2));
  pointer-events: none;
}

.c1:before {
  border-radius: 2rem;
  -webkit-transition: background 0.4s;
  transition: background 0.4s;
}

.c1:after {
  position: absolute;
  display: block;
  content: "";
}

.c1:before {
  position: absolute;
  display: block;
  content: "";
  top: 1px;
  right: 1px;
  bottom: 1px;
  left: 1px;
  background-color: var(--px-toggle__background--unchecked,#d8e0e5);
}

.c1:hover {
  background-color: var(--px-toggle__background-border--unchecked--hover,#677e8c);
}

.c1:hover:before {
  background-color: var(--px-toggle__background--unchecked--hover,#b6c3cc);
}

.c1:active {
  background-color: var(--px-toggle__background-border--unchecked-pressed,#4f6472);
}

.c1:active:before {
  background-color: var(--px-toggle__background--unchecked--pressed,#748b99);
}

.c1:after {
  top: -3px;
  left: -1px;
  height: 1.33333rem;
  width: 1.33333rem;
  border: 1px solid var(--px-toggle__switch-border,#889aa5);
  border-radius: 50%;
  box-shadow: 0 1px 2px var(--px-toggle__switch-shadow-color,rgba(0,0,0,0.15));
  background-color: var(--px-toggle__switch,#fff);
  -webkit-transition: margin 0.4s,background 0.4s,border-color 0.4s;
  transition: margin 0.4s,background 0.4s,border-color 0.4s;
}

.c1:hover:after {
  background-color: var(--px-toggle__switch--hover,#c5d1d8);
  border: 1px solid var(--px-toggle__switch-border--hover,#889aa5);
}

.c1:active:after {
  background-color: var(--px-toggle__switch--pressed,#96a8b2);
  border: 1px solid var(--px-toggle__switch-border--pressed,#889aa5);
}

.c0 {
  position: absolute !important;
  overflow: hidden !important;
  width: 1px !important;
  height: 1px !important;
  margin: -1px !important;
  -webkit-clip: rect(0 0 0 0) !important;
  clip: rect(0 0 0 0) !important;
}

.c0:disabled,
.c0:disabled + label {
  background-color: var(--px-toggle__background-border--unchecked--disabled,rgba(0,0,0,0.2));
  pointer-events: none;
}

.c0:disabled:after,
.c0:disabled + label:after {
  background-color: var(--px-toggle__switch--disabled,#fff);
}

.c0:disabled:before,
.c0:disabled + label:before {
  background-color: var(--px-toggle__background--unchecked--disabled,#fff);
  box-shadow: none;
}

.c0:checked:disabled + label {
  background-color: var(--px-toggle__background--checked--disabled,rgba(0,0,0,0.2));
}

.c0:checked:disabled + label:after {
  box-shadow: none;
  background-color: var(--px-toggle__switch--disabled,white);
  border-width: 1px;
  border-style: solid;
  border-color: var(--px-toggle__switch-border--disabled,rgba(0,0,0,0.2));
}

.c0:checked:disabled + label:before {
  background-color: transparent;
}

.c0:checked + label {
  background-color: var(--px-toggle__background--checked,#007acc);
}

.c0:checked + label:before {
  background-color: var(--px-toggle__background--checked,#007acc);
}

.c0:checked + label:hover {
  background-color: var(--px-toggle__background--checked--hover,#005c99);
}

.c0:checked + label:hover:before {
  background-color: var(--px-toggle__background--checked--hover,#005c99);
}

.c0:checked + label:active {
  background-color: var(--px-toggle__background--checked--pressed,#003d66);
}

.c0:checked + label:active:before {
  background-color: var(--px-toggle__background--checked--pressed,#003d66);
}

.c0:checked + label:after {
  margin-left: 1.467rem;
}

<div
  className={null}
>
  <input
    checked={null}
    className="c0"
    disabled={true}
    id="toggle1"
    name="toggle1"
    size={null}
    type="checkbox"
    value={null}
  />
  <label
    className="c1"
    disabled={true}
    onClick={null}
    size={null}
  />
</div>
`;

exports[`Toggle matches snapshot 1`] = `
.c1 {
  position: relative;
  display: block;
  width: 2.66667rem;
  min-width: 2.66667rem;
  height: calc(1rem - 1px);
  margin-bottom: 0;
  margin-left: 0;
  padding: 1px;
  cursor: pointer;
  outline: 0;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: var(--px-toggle__background-border--unchecked,#889aa5);
  text-indent: -9999rem;
  pointer-events: auto;
  border-radius: 2rem;
  -webkit-transition: background 0.4s;
  transition: background 0.4s;
}

.c1:before {
  border-radius: 2rem;
  -webkit-transition: background 0.4s;
  transition: background 0.4s;
}

.c1:after {
  position: absolute;
  display: block;
  content: "";
}

.c1:before {
  position: absolute;
  display: block;
  content: "";
  top: 1px;
  right: 1px;
  bottom: 1px;
  left: 1px;
  background-color: var(--px-toggle__background--unchecked,#d8e0e5);
}

.c1:hover {
  background-color: var(--px-toggle__background-border--unchecked--hover,#677e8c);
}

.c1:hover:before {
  background-color: var(--px-toggle__background--unchecked--hover,#b6c3cc);
}

.c1:active {
  background-color: var(--px-toggle__background-border--unchecked-pressed,#4f6472);
}

.c1:active:before {
  background-color: var(--px-toggle__background--unchecked--pressed,#748b99);
}

.c1:after {
  top: -3px;
  left: -1px;
  height: 1.33333rem;
  width: 1.33333rem;
  border: 1px solid var(--px-toggle__switch-border,#889aa5);
  border-radius: 50%;
  box-shadow: 0 1px 2px var(--px-toggle__switch-shadow-color,rgba(0,0,0,0.15));
  background-color: var(--px-toggle__switch,#fff);
  -webkit-transition: margin 0.4s,background 0.4s,border-color 0.4s;
  transition: margin 0.4s,background 0.4s,border-color 0.4s;
}

.c1:hover:after {
  background-color: var(--px-toggle__switch--hover,#c5d1d8);
  border: 1px solid var(--px-toggle__switch-border--hover,#889aa5);
}

.c1:active:after {
  background-color: var(--px-toggle__switch--pressed,#96a8b2);
  border: 1px solid var(--px-toggle__switch-border--pressed,#889aa5);
}

.c0 {
  position: absolute !important;
  overflow: hidden !important;
  width: 1px !important;
  height: 1px !important;
  margin: -1px !important;
  -webkit-clip: rect(0 0 0 0) !important;
  clip: rect(0 0 0 0) !important;
}

.c0:disabled,
.c0:disabled + label {
  background-color: var(--px-toggle__background-border--unchecked--disabled,rgba(0,0,0,0.2));
  pointer-events: none;
}

.c0:disabled:after,
.c0:disabled + label:after {
  background-color: var(--px-toggle__switch--disabled,#fff);
}

.c0:disabled:before,
.c0:disabled + label:before {
  background-color: var(--px-toggle__background--unchecked--disabled,#fff);
  box-shadow: none;
}

.c0:checked:disabled + label {
  background-color: var(--px-toggle__background--checked--disabled,rgba(0,0,0,0.2));
}

.c0:checked:disabled + label:after {
  box-shadow: none;
  background-color: var(--px-toggle__switch--disabled,white);
  border-width: 1px;
  border-style: solid;
  border-color: var(--px-toggle__switch-border--disabled,rgba(0,0,0,0.2));
}

.c0:checked:disabled + label:before {
  background-color: transparent;
}

.c0:checked + label {
  background-color: var(--px-toggle__background--checked,#007acc);
}

.c0:checked + label:before {
  background-color: var(--px-toggle__background--checked,#007acc);
}

.c0:checked + label:hover {
  background-color: var(--px-toggle__background--checked--hover,#005c99);
}

.c0:checked + label:hover:before {
  background-color: var(--px-toggle__background--checked--hover,#005c99);
}

.c0:checked + label:active {
  background-color: var(--px-toggle__background--checked--pressed,#003d66);
}

.c0:checked + label:active:before {
  background-color: var(--px-toggle__background--checked--pressed,#003d66);
}

.c0:checked + label:after {
  margin-left: 1.467rem;
}

<div
  className={null}
>
  <input
    checked={null}
    className="c0"
    disabled={null}
    id="toggle1"
    name="toggle1"
    size={null}
    type="checkbox"
    value={null}
  />
  <label
    className="c1"
    disabled={null}
    onClick={null}
    size={null}
  />
</div>
`;
